tests/unit/test_reservation.py

```diff
@@ -208,5 +215,402 @@ def test___heterogenous_session_infos___create_sessions___grouped_by_instrument_
+        assert (
+            "Expected <class 'int'>, got <class 'tests.utilities.fake_driver.Session'>."
+            in exc_info.value.args[0]
+        )
+
+
+def test___session_not_created___get_connection___type_error_raised(
+    session_management_client: Mock,
+) -> None:
+    grpc_session_infos = create_nifake_session_infos(1)
+    grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin1", site=2, channel="3")
+    reservation = MultiSessionReservation(session_management_client, grpc_session_infos)
+
+    with pytest.raises(TypeError) as exc_info:
+        _ = reservation.get_connection(fake_driver.Session)
+
+    assert "Incorrect type for session 'MySession0'." in exc_info.value.args[0]
+    assert (
+        "Expected <class 'tests.utilities.fake_driver.Session'>, got <class 'NoneType'>."
+        in exc_info.value.args[0]
+    )
+
+
+def test___session_not_created___get_connection_with_session_type_object___connection_returned(
+    session_management_client: Mock,
+) -> None:
+    grpc_session_infos = create_nifake_session_infos(1)
+    grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin1", site=2, channel="3")
+    reservation = MultiSessionReservation(session_management_client, grpc_session_infos)
+
+    connection = reservation.get_connection(object)
+
+    assert connection.pin_or_relay_name == "Pin1"
+    assert connection.site == 2
+    assert connection.channel_name == "3"
+    assert connection.session is None
+
+
+def test___reservation_order___get_connections_with_specified_order___connections_returned_in_specified_order(
+    session_management_client: Mock,
+) -> None:
+    grpc_session_infos = _create_grpc_session_infos_for_ordering()
+    reservation = MultiSessionReservation(
+        session_management_client,
+        grpc_session_infos,
+        reserved_pin_or_relay_names=["Pin3", "Pin1", "Pin4", "Pin2"],
+        reserved_sites=[1, 0],
+        reserved_instrument_type_ids=["nifoo", "nibar"],
+    )
+
+    nifoo_connections = reservation.get_connections(
+        object,
+        pin_or_relay_names=["Pin1", "Pin3", "Pin2", "Pin4"],
+        sites=[0, 1],
+        instrument_type_id="nifoo",
+    )
+    nibar_connections = reservation.get_connections(
+        object,
+        pin_or_relay_names=["Pin1", "Pin3", "Pin2", "Pin4"],
+        sites=[0, 1],
+        instrument_type_id="nibar",
+    )
+
+    # Group by instrument_type_id. Sort by site, then by pin name.
+    assert [conn.pin_or_relay_name for conn in nifoo_connections] == [
+        "Pin1",
+        "Pin3",
+        "Pin2",
+        "Pin1",
+        "Pin3",
+        "Pin2",
+    ]
+    assert [conn.site for conn in nifoo_connections] == [0, 0, 0, 1, 1, 1]
+    assert [conn.channel_name for conn in nifoo_connections] == ["0", "2", "1", "3", "5", "4"]
+    assert [conn.pin_or_relay_name for conn in nibar_connections] == ["Pin4", "Pin4"]
+    assert [conn.site for conn in nibar_connections] == [0, 1]
```

Comment:
Building a list of named tuples would make these more readable, compared to the parallel lists.